### PR TITLE
[flakes] add and remove packages, individually

### PR DESCRIPTION
**Pull request by savil**
## Summary

This PR adds the following top-level functions:
- `devbox.addPackagesToProfile`: used during `devbox add` and `devbox shell`.
- `devbox.removePackagesFromProfile`: used during `devbox rm`.

This PR's size comes from inspecting the results of `nix profile list`, which we parse into `nix.ProfileListItem` structs. We use this information in two ways:
1. In `devbox.addPackagesToProfile`, we inspect list of `nix.ProfileListItem` to determine which of the `devbox.json` packages are pending installation.
2. In `devbox.removePackagesFromProfile`, we need the "attribute path" of the package, which we derive from its `nix.ProfileListItem`

UX:
The UX can be improved, but I didn't invest in it since the PR size is already large. That can be a discussion and a follow up PR.

TODOs for future PRs:
- Ensure the `github:NixOS/nixpkgs/<commit>` URL can leverage `nixed` cache service
- During `devbox add` we always list `nixpkgs` as being installed, which is kinda odd. Can we check that it has been pre-fetched already?
- UX improvements
- Older non-flakes profiles link to `bin`, `lib`, `share`, but with flakes profiles we just link to `bin` and `share`

## How was it tested?

```
>  DEVBOX_FEATURE_FLAKES=1 DEVBOX_FEATURE_NIXLESS_SHELL=1 DEVBOX_DEBUG=0 devbox shell
(devbox)>  DEVBOX_FEATURE_FLAKES=1 DEVBOX_FEATURE_NIXLESS_SHELL=1 DEVBOX_DEBUG=0 devbox add vim emacs
(devbox)>  DEVBOX_FEATURE_FLAKES=1 DEVBOX_FEATURE_NIXLESS_SHELL=1 DEVBOX_DEBUG=0 devbox rm emacs
```

Installing looks like below. I did not use the `stepper` component so that we show the output of each individual install.

Screenshot 2023-02-08 at 5 17 39 PM

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #584** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/584" target="_blank">Graphite</a> 
    * **PR #597** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/597" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/597?utm_source=stack-comment).

File: internal/impl/packages.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+	for _, pkg := range d.cfg.Packages {
+		if _, ok := installed[pkg]; !ok {
+			pending = append(pending, pkg)
+		}
+	}
+	return pending, nil
+}
+
+// nixProfileListItems returns a list of the installed packages
+func (d *Devbox) nixProfileListItems() ([]*nixProfileListItem, error) {
```
Comment (savil) [inline internal/impl/packages.go:186]: I think this and below code is better moved to `nix/profile.go` so it'll be `nix.ProfileListItems()`

**Review by gcurtis — APPROVED** [review]
I think package name collisions are the only must-fix thing. Is there a performance difference in invoking Nix for each package vs. doing them all at once?

File: internal/nix/profile.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	parts := strings.Split(attrPath, ".")
+	if len(parts) < 2 {
+		return "", errors.Errorf(
+			"Expected >= 2 parts for AttributePath in nix profile list, but got %d parts. AttributePath: %s",
+			len(parts),
+			attrPath,
+		)
+	}
+
+	packageName := parts[len(parts)-1]
```
[the remark this answers is not in the document]
Comment (savil) [reply internal/nix/profile.go:157]: Good point.

I could parse it the other direction assuming the prefix will always be `legacyPackages.<platform>`

[the remark this answers is not in the document]
Comment (savil) [reply internal/nix/profile.go]: Added a test case and updated code to fix this.

File: internal/nix/profile.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		return nil, errors.WithStack(err)
+	}
+
+	// The `out` output is of the form:
+	// <index> <UnlockedReference> <LockedReference> <NixStorePath>
+	//
+	// Using an example:
+	// 0 github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.go_1_19 github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.go_1_19 /nix/store/w0lyimyyxxfl3gw40n46rpn1yjrl3q85-go-1.19.3
+	// 1 github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.vim github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.vim /nix/store/gapbqxx1d49077jk8ay38z11wgr12p23-vim-9.0.0609
+	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
```
[the remark this answers is not in the document]
Comment (savil) [reply internal/nix/profile.go]: Much nicer now, thanks for the tip to use scanner (and helpful code).

**Pull request merged**